File: 0202-happy-number/0202-happy-number.py
class Solution:
    def isHappy(self, n: int) -> bool:
        dp = set()
        while n not in dp:
            dp.add(n)
            sq = 0
            while n > 0:
                b = n % 10
                sq += b * b
                n //= 10
            if sq == 1: return True
            n = sq
        return False

File: 0202-happy-number/0202-happy-number.cpp
class Solution {
public:
    bool isHappy(int n) {
        std::unordered_set<int> dp;
        while (dp.insert(n).second) {
            int sq = 0;
            while (n > 0) {
                int base = n % 10;
                sq += base * base;
                n /= 10;
            }
            if (sq == 1) {
                return true;
            }
            n = sq;
        }
        return false;
    }
};